overlord/snapstate/handlers.go

```diff
@@ -821,17 +821,38 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) error {
 	return nil
 }
 
-// maybeRestart will schedule a reboot or restart as needed for the just linked
-// snap with info if it's a core or kernel snap.
+// maybeRestart will schedule a reboot or restart as needed for the
+// just linked snap with info if it's a core or snapd or kernel snap.
 func maybeRestart(t *state.Task, info *snap.Info) {
 	st := t.State()
+
+	// TODO: once classic uses the snapd snap we need to restart
+	//       here too
 	if release.OnClassic && info.Type == snap.TypeOS {
 		t.Logf("Requested daemon restart.")
 		st.RequestRestart(state.RestartDaemon)
 	}
-	if !release.OnClassic && boot.KernelOsBaseRebootRequired(info) {
-		t.Logf("Requested system restart.")
-		st.RequestRestart(state.RestartSystem)
+
+	if !release.OnClassic {
```

Comment:
Function logic might be simplified a bit by merging this test in the prior block with: 
```
if release.OnClassic {
    if info.Type == snap.TypeOS {
        ...
    }   
    return
}
```
... then everything else is !classic, which is the tricky case.